library_generation/generate_composed_library.py

```diff
@@ -0,0 +1,139 @@
+  - A "google" folder found in the googleapis repository
+  - A "grafeas" folder found in the googleapis repository
+Note: googleapis repo is found in https://github.com/googleapis/googleapis.
+"""
+
+import click
+import utilities as util
+import os
+import subprocess
+
+script_dir = os.path.dirname(os.path.realpath(__file__))
+
+@click.group(invoke_without_command=False)
+@click.pass_context
+@click.version_option(message="%(version)s")
+def main(ctx):
+    pass
+
+@main.command()
+@click.option(
+    "--generation_queries",
+    required=True,
+    type=str,
+    help="""
+      a single string of comma-separated key-value groups separated by a
+      pipe | (i.e. the groups are spearated by pipe, while a group's key-values are
+      separated by comma). They key-value groups are in the form of `key=value` and will
+      be converted to an argument to generate_library.sh (`--key value`).
+
+      example: "proto_path=google/cloud/asset/v1,destination_path=google-cloud-asset-v1-java,(...)|proto_path=google/cloud/asset/v1p2beta5,destination_path=google-cloud-asset-v1-java,(...)"
+      In this case, generate_library.sh will be called once with
+        generate_library.sh --proto_path google/cloud/asset/v1 --destination_path google-cloud-asset-v1-java
+        and once with
+        generate_library.sh --proto_path google/cloud/asset/v1p2bet5 --destination_path google-cloud-asset-v1p2beta5-java
+    """
+)
+@click.option(
+    "--repository_path",
+    required=True,
+    type=str,
+    help="""
+    A relative path from the output folder (found in cwd when calling this script) that points
+    to the location of the original library. The original library is generally downloaded from
+    github (example, google-cloud-java/java-asset). Postprocessing will be run on such folder,
+    once the pre-processed libraries have been generated
+    """
+)
+@click.option(
+    "--versions_file",
+    required=False,
+    type=str,
+    help="""
+    a file of newline-separated version strings in the form "module:released-version:current-version". 
+    The versions will be applied to the pom.xml files and readmes
+    """
+)
+@click.option(
+    "--enable_postprocessing",
+    required=True,
+    type=str,
+    help="""
+    whether a final postprocessing run should be done on the pre-processed generated libraries
+    """
+)
+def generate(
+    generation_queries,
+    repository_path,
+    versions_file,
+    enable_postprocessing
+):
+  output_folder = util.sh_util('get_output_folder')
+  print(f'output_folder: {output_folder}')
+  os.makedirs(output_folder, exist_ok=True)
+
+  if 'google-cloud-java' in repository_path:
+    print('this is a monorepo library')
+    library = repository_path.split('/')[-1]
+    clone_out = util.sh_util(f'sparse_clone "https://github.com/googleapis/google-cloud-java.git" "{library} google-cloud-pom-parent google-cloud-jar-parent versions.txt .github"', cwd=output_folder)
```

Comment:
Invoking shell command like this is interesting and concise, where did you learn from?

Reply:
I just figured that passing a `list` of arguments (while `source`-ing the utilities script) was too clunky, so I added this util function for when we don't deal with more complex string-with-spaces arguments (because if I pass a `"argument with spaces"` it wont' work.